overlord/configstate/configcore/homedirs_test.go

```diff
@@ -0,0 +1,216 @@
+		if strings.HasPrefix(path, "/home/existingDir") {
+			return &mockedFileInfo{isDir: true}, nil
+		} else if strings.HasPrefix(path, "/home/existingFile") {
+			return &mockedFileInfo{isDir: false}, nil
+		} else {
+			return nil, errors.New("stat failed")
+		}
+	})
+	s.AddCleanup(restore)
+
+}
+
+func (s *homedirsSuite) TestValidationUnhappy(c *C) {
+	for _, testData := range []struct {
+		homedirs      string
+		expectedError string
+	}{
+		{"./here", `path "\./here" is not absolute`},
+		// empty path in list
+		{",/home", `path "" is not absolute`},
+		{"/somewhere/else", `path "/somewhere/else/" unsupported: must start with one of: /home/`},
+		{"/home/foo[12]", `home path invalid: "/home/foo\[12\]" contains a reserved apparmor char.*`},
+		{"/lib/homes", `path "/lib/homes/" uses reserved root directory "/lib/"`},
+		{"/home/missing", `cannot get directory info for "/home/missing/".*`},
+		{"/home/existingFile", `path "/home/existingFile/" is not a directory`},
+	} {
+		err := configcore.Run(coreDev, &mockConf{
+			state: s.state,
+			conf: map[string]interface{}{
+				"homedirs": testData.homedirs,
+			},
+		})
+		c.Assert(err, ErrorMatches, testData.expectedError, Commentf("%v", testData.homedirs))
+	}
+}
+
+func (s *homedirsSuite) TestConfigureOpenFailure(c *C) {
+	var systemInfoPath string
+	restore := configcore.MockOpenFile(func(path string, flags int, mode os.FileMode) (*os.File, error) {
+		systemInfoPath = path
+		return nil, errors.New("open failure")
+	})
+	defer restore()
+	err := configcore.Run(coreDev, &mockConf{
+		state: s.state,
+		conf: map[string]interface{}{
+			"homedirs": "/home/existingDir",
+		},
+	})
+	expectedPath := filepath.Join(dirs.SnapdStateDir(dirs.GlobalRootDir), "system-params")
+	c.Check(systemInfoPath, Equals, expectedPath)
+	c.Assert(err, ErrorMatches, "open failure")
+}
+
+func (s *homedirsSuite) TestConfigureWriteFailure(c *C) {
+	restore := configcore.MockOpenFile(func(path string, flags int, mode os.FileMode) (*os.File, error) {
+		file, err := os.OpenFile(path, flags, mode)
+		// But now close the file so that the write will fail
+		file.Close()
+		return file, err
+	})
+	defer restore()
+	err := configcore.Run(coreDev, &mockConf{
+		state: s.state,
+		conf: map[string]interface{}{
+			"homedirs": "/home/existingDir",
+		},
+	})
+	c.Assert(err, ErrorMatches, "write .*system-params: file already closed")
+}
+
+func (s *homedirsSuite) TestConfigureApparmorTunableFailure(c *C) {
+	var homedirs []string
+	restore := configcore.MockApparmorUpdateHomedirsTunable(func(paths []string) error {
+		homedirs = paths
+		return errors.New("tunable error")
+	})
+	defer restore()
+	err := configcore.Run(coreDev, &mockConf{
+		state: s.state,
+		conf: map[string]interface{}{
+			"homedirs": "/home/existingDir/one,/home/existingDir/two",
+		},
+	})
+	c.Check(homedirs, DeepEquals, []string{"/home/existingDir/one", "/home/existingDir/two"})
+	c.Assert(err, ErrorMatches, "tunable error")
```

Comment:
(nitpick^9) same here as above, I personally would check the error first

Reply:
updated